<a href="https://colab.research.google.com/github/estebanszd/Clase-Introducci-n-/blob/main/Parcial_An%C3%A1lisis_Esteban_szd_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1594]:
#Importamos las librerias que nos van a ayudar a limpiar y organizar el dataframe
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#**Nivel 1: Diagnostico**
##Objetivo: construir un inventario preciso del estado del dataset antes de cualquier intervención.


In [1595]:
#Importamos el dataframe con el pd.read_csv y usamos el path para añadirlo
df = pd.read_csv('/content/prescripciones_raw_suarez_diaz.csv')

In [1596]:
#Para ver el tamaño del dataframe utilizamos .shape y así vemos (FILAS, COLUMNAS)
df.shape

(520, 10)

In [1597]:
#Para ver la información del dataset(Nombres de columnas, Tipo de dato, Espacios vacíos, Uso de memoria)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 520 entries, 0 to 519
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_paciente         520 non-null    int64  
 1   fecha_prescripcion  520 non-null    object 
 2   medicamento         520 non-null    object 
 3   dosis_mg            488 non-null    object 
 4   via_administracion  520 non-null    object 
 5   diagnostico_cie10   401 non-null    object 
 6   peso_paciente_kg    499 non-null    float64
 7   edad                505 non-null    float64
 8   eps                 520 non-null    object 
 9   prescriptor_id      475 non-null    object 
dtypes: float64(2), int64(1), object(7)
memory usage: 40.8+ KB


In [1598]:
#Con el describe podemos ver datos de las variables numéricas, tanto minimo como máximo, promedio, todo aquello que nos ayuda a entender un poc más el dataframe e identificar extremos
df.describe()

,id_paciente,peso_paciente_kg,edad
count,520.000000,499.000000,505.000000
mean,5489.478846,68.670577,51.219802
std,2609.767333,28.232942,21.242113
min,1055.000000,2.000000,-5.000000
25%,3167.500000,59.841224,35.000000
50%,5681.500000,67.971325,51.000000
75%,7683.750000,76.447961,68.000000
max,9975.000000,450.000000,200.000000


In [1599]:
#Con el .dtypes podemos ver el tipo de variable
df.dtypes

,0
id_paciente,int64
fecha_prescripcion,object
medicamento,object
dosis_mg,object
via_administracion,object
diagnostico_cie10,object
peso_paciente_kg,float64
edad,float64
eps,object
prescriptor_id,object


In [1600]:
#Imprimimos el encabezado de el dataframe para hacernos una idea de como es
df.head(7)

,id_paciente,fecha_prescripcion,medicamento,dosis_mg,via_administracion,diagnostico_cie10,peso_paciente_kg,edad,eps,prescriptor_id
0,3118,25/10/2024,METFORMINA,200.0,SC,J06.9,57.964340,37.0,SANITAS,MED-8902
1,4108,2023-06-16,metformina,400.0,via oral,J06.9,61.531964,38.0,nueva eps,MED-3356
2,9375,2026-03-12,ASA,850.0,Oral,K29.5,61.452613,26.0,Sura,MED-9646
3,2171,2023-09-02,Enalapril,200.0,subcutanea,NaN,79.590069,81.0,Nueva EPS,MED-0072
4,6722,05/07/2025,Metformina 500,400.0,Intravenosa,hipertension,65.768793,70.0,Compensar,MED-3734
5,5997,2025-08-01,Ibuprofeno,400.0,Oral,J18.9,65.354706,62.0,Compensar,MED-5965
6,2095,junio 26 2025,amoxicilina,250.0,SC,NaN,48.502784,42.0,EPS Sura,MED-7370


In [1601]:
#Con el .isnull() podemos saber cuando son valores faltantes y el .sum() nos ayuda a saber la totalidad de valores faltantes en el dataframe
nulos = df.isnull().sum()

#Imprimimos para ver la cantidad de valores faltantes por columna
print(nulos)

id_paciente             0
fecha_prescripcion      0
medicamento             0
dosis_mg               32
via_administracion      0
diagnostico_cie10     119
peso_paciente_kg       21
edad                   15
eps                     0
prescriptor_id         45
dtype: int64


In [1602]:
#El porcentaje de valores faltantes lo tomamos por el largo del data frame que divide la totalida de valores faltantes * 100
porcentaje = (df.isnull().sum()/ len(df))*100

#Imprimimos para ver el porcentaje de valores faltantes por columna
print(porcentaje)

id_paciente            0.000000
fecha_prescripcion     0.000000
medicamento            0.000000
dosis_mg               6.153846
via_administracion     0.000000
diagnostico_cie10     22.884615
peso_paciente_kg       4.038462
edad                   2.884615
eps                    0.000000
prescriptor_id         8.653846
dtype: float64


In [1603]:
#Creamos el dataframe con pc.dataframe y a cada columna le damos sus valores, en este caso nulos y porcentaje
df_faltantes = pd.DataFrame({'Valores Faltantes':nulos,
                        '%  De Valores Faltantes':porcentaje})

#Imprimimos el dataframe de valores faltantes
df_faltantes

,Valores Faltantes,% De Valores Faltantes
id_paciente,0,0.000000
fecha_prescripcion,0,0.000000
medicamento,0,0.000000
dosis_mg,32,6.153846
via_administracion,0,0.000000
diagnostico_cie10,119,22.884615
peso_paciente_kg,21,4.038462
edad,15,2.884615
eps,0,0.000000
prescriptor_id,45,8.653846


In [1604]:
#Para hacer que veamos el dataframe de forma descendete utilizamos .sortvalues y en by= (La columna que tomamos en cuenta los valores) y ascending = False para que sea de menos a mayor
df_faltantes_descendente = df_faltantes.sort_values(by='%  De Valores Faltantes', ascending=False)

#Imprimos el dataframe descendente para verificar que quedó bien
df_faltantes_descendente

,Valores Faltantes,% De Valores Faltantes
diagnostico_cie10,119,22.884615
prescriptor_id,45,8.653846
dosis_mg,32,6.153846
peso_paciente_kg,21,4.038462
edad,15,2.884615
id_paciente,0,0.000000
fecha_prescripcion,0,0.000000
medicamento,0,0.000000
via_administracion,0,0.000000
eps,0,0.000000


In [1605]:
#Imprimimos el unique de medicamento y eps para saber que cosas aparecen en cada columna
print(df['medicamento'].unique())

print(df['eps'].unique())


['METFORMINA' 'metformina' 'ASA' 'Enalapril' 'Metformina 500'
 'Ibuprofeno ' 'amoxicilina' 'ENALAPRIL' 'Aspirina ' 'ibuprofeno'
 'Amoxicilina']
['SANITAS' 'nueva eps' 'Sura' 'Nueva EPS' 'Compensar' 'EPS Sura'
 'EPS Sanitas' 'NuevaEPS' 'SURA' 'Sanitas' 'COMPENSAR']


# **Documentación de problemas**
| Columna | Tipo de problema | Ejemplo concreto |
|---|---|---|
| id_paciente | Duplicados parciales (mismo paciente, distintas fechas/medicamentos) | id 5500 aparece varias veces |
| fecha_prescripcion | Inconsistencia de formato: tres formatos distintos | '25/10/2024', '2023-06-16', 'junio 26 2025' |
| medicamento | Variaciones en mayúsculas, espacios y abreviaciones | 'METFORMINA', 'metformina', 'Metformina 500', 'ASA' |
| dosis_mg | Tipo incorrecto (string), strings mixtos con unidad, valores negativos, NaN | '500mg', '-250.0', NaN |
| via_administracion | Inconsistencia categórica: múltiples formas para la misma vía | 'SC', 'Subcutanea', 'subcutanea'; 'VO', 'oral', 'Oral', 'via oral' |
| diagnostico_cie10 | Valores faltantes (22.88% NaN) | NaN en 119 filas |
| peso_paciente_kg | Outliers fisiológicamente imposibles, NaN | 2.0 kg, 450.0 kg; 21 NaN |
| edad | Valores negativos, valor imposible alto, NaN | -5.0, 200.0; 15 NaN |
| eps | Inconsistencia categórica: múltiples variantes para la misma entidad | 'SANITAS', 'Sanitas', 'EPS Sanitas'; 'Sura', 'SURA', 'EPS Sura' |
| prescriptor_id | Valores faltantes (8.65% NaN) | NaN en 45 filas |

#**Nivel 2: Limpieza**
##Objetivo: transformar cada columna al estado que permite análisis confiable.

In [1606]:
# Contamos los valores nulos en la columna 'fecha_prescripcion' antes de la conversión.
nulos_fecha = df['fecha_prescripcion'].isna().sum()

print(nulos_fecha)

0


In [1607]:
# Convertimos la columna 'fecha_prescripcion' a formato datetime, manejando diferentes formatos y errores.
df['fecha_prescripcion'] = pd.to_datetime(
    df['fecha_prescripcion'],
    dayfirst=True,
    format='mixed',
    errors='coerce'
)

In [1608]:
# Verificamos nuevamente la cantidad de nulos en 'fecha_prescripcion' después de la conversión.
nulos_fecha = df['fecha_prescripcion'].isna().sum()

print(nulos_fecha)

77


In [1609]:
# Calculamos la mediana de la columna 'fecha_prescripcion' para rellenar los valores nulos.
mediana_fecha = df['fecha_prescripcion'].dropna().sort_values().iloc[len(df['fecha_prescripcion'].dropna()) // 2]

# Rellenamos los NaT (Not a Time) con la mediana calculada.
df['fecha_prescripcion'] = df['fecha_prescripcion'].fillna(mediana_fecha)

In [1610]:
# Verificamos que no queden valores nulos en la columna 'fecha_prescripcion' después de rellenar.
nulos_fecha = df['fecha_prescripcion'].isna().sum()

print(nulos_fecha)

0


In [1611]:
# Mostramos las primeras filas del DataFrame para ver los cambios realizados en las fechas.
df.head()

,id_paciente,fecha_prescripcion,medicamento,dosis_mg,via_administracion,diagnostico_cie10,peso_paciente_kg,edad,eps,prescriptor_id
0,3118,2024-10-25,METFORMINA,200.0,SC,J06.9,57.964340,37.0,SANITAS,MED-8902
1,4108,2023-06-16,metformina,400.0,via oral,J06.9,61.531964,38.0,nueva eps,MED-3356
2,9375,2026-03-12,ASA,850.0,Oral,K29.5,61.452613,26.0,Sura,MED-9646
3,2171,2023-09-02,Enalapril,200.0,subcutanea,NaN,79.590069,81.0,Nueva EPS,MED-0072
4,6722,2025-07-05,Metformina 500,400.0,Intravenosa,hipertension,65.768793,70.0,Compensar,MED-3734


In [1612]:
# Normalizamos los campos 'medicamento' y 'via_administracion': eliminamos espacios y convertimos a minúsculas.
df['medicamento'] = df['medicamento'].str.strip().str.lower()
df['via_administracion'] = df['via_administracion'].str.strip().str.lower()

In [1613]:
# Mostramos las primeras filas del DataFrame para verificar la normalización de texto.
df.head()

,id_paciente,fecha_prescripcion,medicamento,dosis_mg,via_administracion,diagnostico_cie10,peso_paciente_kg,edad,eps,prescriptor_id
0,3118,2024-10-25,metformina,200.0,sc,J06.9,57.964340,37.0,SANITAS,MED-8902
1,4108,2023-06-16,metformina,400.0,via oral,J06.9,61.531964,38.0,nueva eps,MED-3356
2,9375,2026-03-12,asa,850.0,oral,K29.5,61.452613,26.0,Sura,MED-9646
3,2171,2023-09-02,enalapril,200.0,subcutanea,NaN,79.590069,81.0,Nueva EPS,MED-0072
4,6722,2025-07-05,metformina 500,400.0,intravenosa,hipertension,65.768793,70.0,Compensar,MED-3734


In [1614]:
# Creamos diccionarios de mapeo para unificar variantes de 'vias de administración' y 'medicamentos'.
mapeo_vias = {
    'via oral': 'oral',
    'vo': 'oral',
    'iv': 'intravenosa',
    'subcutanea': 'subcutanea',
    'sc': 'subcutanea',
    'oral': 'oral',
    'intravenosa': 'intravenosa'
}
mapeo_medicamentos = {
    'asa': 'aspirina',
    'aspirina': 'aspirina',
    'metformina 500': 'metformina',
    'metformina': 'metformina',
    'ibuprofeno': 'ibuprofeno',
    'enalapril': 'enalapril',
    'losartan': 'losartan'
}

In [1615]:
# Aplicamos los mapeos para estandarizar los valores en las columnas 'via_administracion' y 'medicamento'.
df['via_administracion'] = df['via_administracion'].replace(mapeo_vias)
df['medicamento'] = df['medicamento'].replace(mapeo_medicamentos)

In [1616]:
# Imprimimos los valores únicos de 'via_administracion' y 'medicamento' para verificar la unificación.
print(df['via_administracion'].unique())

print(df['medicamento'].unique())

['subcutanea' 'oral' 'intravenosa']
['metformina' 'aspirina' 'enalapril' 'ibuprofeno' 'amoxicilina']


In [1617]:
# Mostramos las primeras 20 filas del DataFrame para observar los cambios en las columnas unificadas.
df.head(20)

,id_paciente,fecha_prescripcion,medicamento,dosis_mg,via_administracion,diagnostico_cie10,peso_paciente_kg,edad,eps,prescriptor_id
0,3118,2024-10-25,metformina,200.0,subcutanea,J06.9,57.964340,37.0,SANITAS,MED-8902
1,4108,2023-06-16,metformina,400.0,oral,J06.9,61.531964,38.0,nueva eps,MED-3356
2,9375,2026-03-12,aspirina,850.0,oral,K29.5,61.452613,26.0,Sura,MED-9646
3,2171,2023-09-02,enalapril,200.0,subcutanea,NaN,79.590069,81.0,Nueva EPS,MED-0072
4,6722,2025-07-05,metformina,400.0,intravenosa,hipertension,65.768793,70.0,Compensar,MED-3734
5,5997,2025-08-01,ibuprofeno,400.0,oral,J18.9,65.354706,62.0,Compensar,MED-5965
6,2095,2024-11-15,amoxicilina,250.0,subcutanea,NaN,48.502784,42.0,EPS Sura,MED-7370
7,7997,2025-04-28,metformina,250.0,subcutanea,NaN,60.064021,41.0,EPS Sanitas,MED-9616
8,5682,2024-07-27,amoxicilina,850.0,oral,diabetes tipo 2,55.679515,59.0,NuevaEPS,MED-2351
9,3018,2023-06-08,enalapril,500mg,intravenosa,diabetes tipo 2,84.012844,43.0,SURA,MED-7564


In [1618]:
# Imprimimos la cantidad de variantes únicas en la columna 'eps' antes de la unificación y su recuento.
print('Variantes únicas de EPS antes de unificar:', df['eps'].nunique())
print(df['eps'].value_counts())

Variantes únicas de EPS antes de unificar: 11
eps
Sura           58
Compensar      54
SURA           53
NuevaEPS       50
SANITAS        50
nueva eps      49
Sanitas        46
Nueva EPS      44
EPS Sanitas    44
EPS Sura       36
COMPENSAR      36
Name: count, dtype: int64


In [1619]:
# Normalizamos la columna 'eps' convirtiendo a minúsculas y eliminando espacios.
df['eps'] = df['eps'].str.strip().str.lower()

# Creamos un diccionario de mapeo para unificar las diferentes representaciones de las EPS.
mapeo_eps = {
    'sanitas': 'eps sanitas',
    'eps sanitas': 'eps sanitas',
    'nueva eps': 'nueva eps',
    'nuevaeps': 'nueva eps',
    'sura': 'eps sura',
    'eps sura': 'eps sura',
    'compensar': 'compensar eps',
    'eps compensar': 'compensar eps',
    'salud total': 'salud total'
}

# Aplicamos el mapeo a la columna 'eps'.
df['eps'] = df['eps'].replace(mapeo_eps)

In [1620]:
# Imprimimos la cantidad de variantes únicas de 'eps' después de la unificación y su recuento.
print('Variantes únicas de eps despues de unificar:', df['eps'].nunique())
print(df['eps'].value_counts())

Variantes únicas de eps despues de unificar: 4
eps
eps sura         147
nueva eps        143
eps sanitas      140
compensar eps     90
Name: count, dtype: int64


In [1621]:
# Mostramos las primeras 20 filas del DataFrame para verificar los cambios en la columna 'eps'.
df.head(20)

,id_paciente,fecha_prescripcion,medicamento,dosis_mg,via_administracion,diagnostico_cie10,peso_paciente_kg,edad,eps,prescriptor_id
0,3118,2024-10-25,metformina,200.0,subcutanea,J06.9,57.964340,37.0,eps sanitas,MED-8902
1,4108,2023-06-16,metformina,400.0,oral,J06.9,61.531964,38.0,nueva eps,MED-3356
2,9375,2026-03-12,aspirina,850.0,oral,K29.5,61.452613,26.0,eps sura,MED-9646
3,2171,2023-09-02,enalapril,200.0,subcutanea,NaN,79.590069,81.0,nueva eps,MED-0072
4,6722,2025-07-05,metformina,400.0,intravenosa,hipertension,65.768793,70.0,compensar eps,MED-3734
5,5997,2025-08-01,ibuprofeno,400.0,oral,J18.9,65.354706,62.0,compensar eps,MED-5965
6,2095,2024-11-15,amoxicilina,250.0,subcutanea,NaN,48.502784,42.0,eps sura,MED-7370
7,7997,2025-04-28,metformina,250.0,subcutanea,NaN,60.064021,41.0,eps sanitas,MED-9616
8,5682,2024-07-27,amoxicilina,850.0,oral,diabetes tipo 2,55.679515,59.0,nueva eps,MED-2351
9,3018,2023-06-08,enalapril,500mg,intravenosa,diabetes tipo 2,84.012844,43.0,eps sura,MED-7564


In [1622]:
# Mostramos estadísticas descriptivas para la columna 'peso_paciente_kg'.
print(df['peso_paciente_kg'].describe())

count    499.000000
mean      68.670577
std       28.232942
min        2.000000
25%       59.841224
50%       67.971325
75%       76.447961
max      450.000000
Name: peso_paciente_kg, dtype: float64


In [1623]:
# Calculamos el primer cuartil (Q1), tercer cuartil (Q3) y el rango intercuartílico (IQR) para 'peso_paciente_kg'.
Q1 = df['peso_paciente_kg'].quantile(0.25)
Q3 = df['peso_paciente_kg'].quantile(0.75)
IQR = Q3 - Q1

# Calculamos los límites inferior y superior para identificar outliers utilizando el método IQR.
limite_inferior = Q1 - 1.5 * IQR
limite_superior = Q3 + 1.5 * IQR

# Imprimimos los valores calculados para análisis.
print(f'Q1: {Q1:.2f} | Q3: {Q3:.2f} | IQR: {IQR:.2f}')
print(f'Límite inferior: {limite_inferior:.2f} | Límite superior: {limite_superior:.2f}')

Q1: 59.84 | Q3: 76.45 | IQR: 16.61
Límite inferior: 34.93 | Límite superior: 101.36


In [1624]:
# Identificamos los outliers en 'peso_paciente_kg' basándonos en los límites calculados.
outliers = df[(df['peso_paciente_kg'] < limite_inferior) | (df['peso_paciente_kg'] > limite_superior)]

# Imprimimos la cantidad de outliers detectados.
print(f'\nOutliers detectados: {len(outliers)}')


Outliers detectados: 13


In [1625]:
# Calculamos la mediana de 'peso_paciente_kg' para rellenar los outliers y valores NaN.
mediana_peso = df['peso_paciente_kg'].median()

# Reemplazamos los outliers por la mediana.
df.loc[
    (df['peso_paciente_kg'] < limite_inferior) | (df['peso_paciente_kg'] > limite_superior),
    'peso_paciente_kg'
] = mediana_peso

# También rellenamos los NaN con la mediana.
df['peso_paciente_kg'] = df['peso_paciente_kg'].fillna(mediana_peso)

# Mostramos las estadísticas descriptivas de 'peso_paciente_kg' después del rellenado.
print(df['peso_paciente_kg'].describe())

count    520.000000
mean      68.109038
std       11.307233
min       37.176289
25%       61.134007
50%       67.971325
75%       75.931387
max       98.098533
Name: peso_paciente_kg, dtype: float64


In [1626]:
# Mostramos las primeras 15 filas del DataFrame para verificar los cambios en 'peso_paciente_kg'.
df.head(15)

,id_paciente,fecha_prescripcion,medicamento,dosis_mg,via_administracion,diagnostico_cie10,peso_paciente_kg,edad,eps,prescriptor_id
0,3118,2024-10-25,metformina,200.0,subcutanea,J06.9,57.964340,37.0,eps sanitas,MED-8902
1,4108,2023-06-16,metformina,400.0,oral,J06.9,61.531964,38.0,nueva eps,MED-3356
2,9375,2026-03-12,aspirina,850.0,oral,K29.5,61.452613,26.0,eps sura,MED-9646
3,2171,2023-09-02,enalapril,200.0,subcutanea,NaN,79.590069,81.0,nueva eps,MED-0072
4,6722,2025-07-05,metformina,400.0,intravenosa,hipertension,65.768793,70.0,compensar eps,MED-3734
5,5997,2025-08-01,ibuprofeno,400.0,oral,J18.9,65.354706,62.0,compensar eps,MED-5965
6,2095,2024-11-15,amoxicilina,250.0,subcutanea,NaN,48.502784,42.0,eps sura,MED-7370
7,7997,2025-04-28,metformina,250.0,subcutanea,NaN,60.064021,41.0,eps sanitas,MED-9616
8,5682,2024-07-27,amoxicilina,850.0,oral,diabetes tipo 2,55.679515,59.0,nueva eps,MED-2351
9,3018,2023-06-08,enalapril,500mg,intravenosa,diabetes tipo 2,84.012844,43.0,eps sura,MED-7564


In [1627]:
# Mostramos estadísticas descriptivas para la columna 'edad'.
print(df['edad'].describe())

count    505.000000
mean      51.219802
std       21.242113
min       -5.000000
25%       35.000000
50%       51.000000
75%       68.000000
max      200.000000
Name: edad, dtype: float64


In [1628]:
# Reemplazamos los valores de 'edad' menores a 0 por NaN, ya que no existe nadie menor de 0 años.
df.loc[df['edad'] < 0, 'edad'] = np.nan

In [1629]:
# Calculamos Q1, Q3 e IQR para la columna 'edad'.
Q1_edad = df['edad'].quantile(0.25)
Q3_edad = df['edad'].quantile(0.75)
IQR_edad = Q3_edad - Q1_edad

# Calculamos los límites inferior y superior para detectar outliers en 'edad', asegurando que el límite inferior no sea negativo.
limite_inf_edad = max(0, Q1_edad - 1.5 * IQR_edad)
limite_sup_edad = Q3_edad + 1.5 * IQR_edad

# Imprimimos los valores calculados.
print(f'Q1: {Q1_edad} | Q3: {Q3_edad} | IQR: {IQR_edad}')
print(f'Límite inferior: {limite_inf_edad:.2f} | Límite superior: {limite_sup_edad:.2f}')

# Identificamos los outliers en 'edad'.
outliers_edad = df[(df['edad'] < limite_inf_edad) | (df['edad'] > limite_sup_edad)]
print(f'\nOutliers de edad detectados: {len(outliers_edad)}')
print('Valores:', sorted(outliers_edad['edad'].dropna().unique()))

Q1: 35.25 | Q3: 68.0 | IQR: 32.75
Límite inferior: 0.00 | Límite superior: 117.12

Outliers de edad detectados: 1
Valores: [np.float64(200.0)]


In [1630]:
# Calculamos la mediana de 'edad' para rellenar los outliers y valores NaN.
mediana_edad = df['edad'].median()

# Reemplazamos los outliers en 'edad' por la mediana.
df.loc[
    (df['edad'] < limite_inf_edad) | (df['edad'] > limite_sup_edad),
    'edad'
] = mediana_edad
# Rellenamos cualquier valor NaN restante en 'edad' con la mediana.
df['edad'] = df['edad'].fillna(mediana_edad)
#Volvemos a edad un entero para que no queden decimales en edad
df['edad'] = df['edad'].astype(int)

In [1631]:
df.head()

,id_paciente,fecha_prescripcion,medicamento,dosis_mg,via_administracion,diagnostico_cie10,peso_paciente_kg,edad,eps,prescriptor_id
0,3118,2024-10-25,metformina,200.0,subcutanea,J06.9,57.964340,37,eps sanitas,MED-8902
1,4108,2023-06-16,metformina,400.0,oral,J06.9,61.531964,38,nueva eps,MED-3356
2,9375,2026-03-12,aspirina,850.0,oral,K29.5,61.452613,26,eps sura,MED-9646
3,2171,2023-09-02,enalapril,200.0,subcutanea,NaN,79.590069,81,nueva eps,MED-0072
4,6722,2025-07-05,metformina,400.0,intravenosa,hipertension,65.768793,70,compensar eps,MED-3734


In [1632]:
# Mostramos estadísticas descriptivas para la columna 'edad' después de la limpieza y rellenado.
print(df['edad'].describe())

count    520.000000
mean      51.657692
std       18.847378
min       18.000000
25%       36.750000
50%       51.000000
75%       68.000000
max       84.000000
Name: edad, dtype: float64


In [1633]:
# Usamos expresiones regulares para extraer solo los números (incluyendo decimales)
df['dosis_mg'] = df['dosis_mg'].astype(str).str.extract(r'(\d+\.?\d*)').astype(float)

# Convertimos negativos a NaN para que sean imputados en el siguiente paso
df.loc[df['dosis_mg'] < 0, 'dosis_mg'] = np.nan

# Convertimos valores mayores de 2000 ya que no hay paciente que reciba más de 1.000 a excepción de unos outliers que aparecián con 50.000
df.loc[df['dosis_mg'] > 2000, 'dosis_mg'] = np.nan

# Usamos transform para mantener el tamaño original del dataframe
df['dosis_mg'] = df['dosis_mg'].fillna(
    df.groupby('medicamento')['dosis_mg'].transform('median')
)

In [1634]:
# Mostramos las primeras 20 filas del DataFrame para verificar los cambios en 'dosis_mg'.
df.head(20)

,id_paciente,fecha_prescripcion,medicamento,dosis_mg,via_administracion,diagnostico_cie10,peso_paciente_kg,edad,eps,prescriptor_id
0,3118,2024-10-25,metformina,200.0,subcutanea,J06.9,57.964340,37,eps sanitas,MED-8902
1,4108,2023-06-16,metformina,400.0,oral,J06.9,61.531964,38,nueva eps,MED-3356
2,9375,2026-03-12,aspirina,850.0,oral,K29.5,61.452613,26,eps sura,MED-9646
3,2171,2023-09-02,enalapril,200.0,subcutanea,NaN,79.590069,81,nueva eps,MED-0072
4,6722,2025-07-05,metformina,400.0,intravenosa,hipertension,65.768793,70,compensar eps,MED-3734
5,5997,2025-08-01,ibuprofeno,400.0,oral,J18.9,65.354706,62,compensar eps,MED-5965
6,2095,2024-11-15,amoxicilina,250.0,subcutanea,NaN,48.502784,42,eps sura,MED-7370
7,7997,2025-04-28,metformina,250.0,subcutanea,NaN,60.064021,41,eps sanitas,MED-9616
8,5682,2024-07-27,amoxicilina,850.0,oral,diabetes tipo 2,55.679515,59,nueva eps,MED-2351
9,3018,2023-06-08,enalapril,500.0,intravenosa,diabetes tipo 2,84.012844,43,eps sura,MED-7564


In [1635]:
# miramos elnúmero total de filas para saber cual fue el cambio.
len(df)

520

In [1636]:
# Guardamos el número de filas antes de eliminar duplicados exactos.
n_antes = len(df)
# Eliminamos filas duplicadas exactas de todo el DataFrame.
df = df.drop_duplicates()
# Calculamos cuántos duplicados exactos fueron eliminados.
n_exactos = n_antes - len(df)
print(f'Duplicados exactos eliminados: {n_exactos}')
print(f'Filas restantes: {len(df)}')

Duplicados exactos eliminados: 20
Filas restantes: 500


In [1637]:
#Identificamos y eliminamos duplicados parciales: mismo id_paciente + fecha_prescripcion + medicamento
n_antes = len(df)
df = df.drop_duplicates(subset=['id_paciente', 'fecha_prescripcion', 'medicamento'], keep='first')
n_parciales = n_antes - len(df)
print(f'Duplicados parciales eliminados: {n_parciales}')
print(f'Filas restantes tras limpieza de duplicados: {len(df)}')

Duplicados parciales eliminados: 0
Filas restantes tras limpieza de duplicados: 500


In [1638]:
# Mostramos las primeras 20 filas del DataFrame después de la eliminación de duplicados.
df.head(20)

,id_paciente,fecha_prescripcion,medicamento,dosis_mg,via_administracion,diagnostico_cie10,peso_paciente_kg,edad,eps,prescriptor_id
0,3118,2024-10-25,metformina,200.0,subcutanea,J06.9,57.964340,37,eps sanitas,MED-8902
1,4108,2023-06-16,metformina,400.0,oral,J06.9,61.531964,38,nueva eps,MED-3356
2,9375,2026-03-12,aspirina,850.0,oral,K29.5,61.452613,26,eps sura,MED-9646
3,2171,2023-09-02,enalapril,200.0,subcutanea,NaN,79.590069,81,nueva eps,MED-0072
4,6722,2025-07-05,metformina,400.0,intravenosa,hipertension,65.768793,70,compensar eps,MED-3734
5,5997,2025-08-01,ibuprofeno,400.0,oral,J18.9,65.354706,62,compensar eps,MED-5965
6,2095,2024-11-15,amoxicilina,250.0,subcutanea,NaN,48.502784,42,eps sura,MED-7370
7,7997,2025-04-28,metformina,250.0,subcutanea,NaN,60.064021,41,eps sanitas,MED-9616
8,5682,2024-07-27,amoxicilina,850.0,oral,diabetes tipo 2,55.679515,59,nueva eps,MED-2351
9,3018,2023-06-08,enalapril,500.0,intravenosa,diabetes tipo 2,84.012844,43,eps sura,MED-7564


#**Nivel 3: Transformación y validación**
##Objetivo: generar variables derivadas y garantizar la integridad del dataset final.

In [1639]:
# Definimos una función para clasificar la edad en rangos etarios.
def clasificar_edad(edad):
    if pd.isna(edad):
        return 'desconocido'
    elif edad < 18:
        return 'pediátrico'
    elif edad <= 64:
        return 'adulto'
    else:
        return 'adulto mayor'

In [1640]:
# Aplicamos la función 'clasificar_edad' para crear una nueva columna 'rango_etario'.
df['rango_etario'] = df['edad'].apply(clasificar_edad)

# Imprimimos la distribución de los rangos etarios.
print('Distribución de rango_etario:')
print(df['rango_etario'].value_counts())

Distribución de rango_etario:
rango_etario
adulto          347
adulto mayor    153
Name: count, dtype: int64


In [1641]:
# Aplicamos la función 'clasificar_edad' para crear una nueva columna 'rango_etario'.
df['rango_etario'] = df['edad'].apply(clasificar_edad)

# Imprimimos la distribución de los rangos etarios.
print('Distribución de rango_etario:')
print(df['rango_etario'].value_counts())

Distribución de rango_etario:
rango_etario
adulto          347
adulto mayor    153
Name: count, dtype: int64


In [1642]:
# Creamos una tabla resumen agrupando por EPS y Vía de Administración, calculando estadísticas de 'dosis_mg'.
tabla_resumen = df.groupby(['eps', 'via_administracion'])['dosis_mg'].agg(
    conteo_dosis='count',
    media_dosis='mean',
    mediana_dosis='median',
    desviacion_std_dosis='std',
    minimo_dosis='min',
    maximo_dosis='max'
).round(2).reset_index()
#Imprimoos tabla de resumen
print('Tabla resumen — Prescripciones por EPS y Vía de Administración:')
tabla_resumen

Tabla resumen — Prescripciones por EPS y Vía de Administración:


,eps,via_administracion,conteo_dosis,media_dosis,mediana_dosis,desviacion_std_dosis,minimo_dosis,maximo_dosis
0,compensar eps,intravenosa,26,561.54,425.0,324.13,100.0,1000.0
1,compensar eps,oral,34,457.35,400.0,305.29,100.0,1000.0
2,compensar eps,subcutanea,27,483.33,400.0,285.21,100.0,1000.0
3,eps sanitas,intravenosa,40,435.00,400.0,266.31,100.0,1000.0
4,eps sanitas,oral,53,503.77,400.0,319.53,100.0,1000.0
5,eps sanitas,subcutanea,40,510.00,400.0,343.66,100.0,1000.0
6,eps sura,intravenosa,48,484.38,400.0,304.60,100.0,1000.0
7,eps sura,oral,55,516.36,400.0,340.76,100.0,1000.0
8,eps sura,subcutanea,42,402.38,400.0,272.96,100.0,1000.0
9,nueva eps,intravenosa,35,468.57,400.0,299.28,100.0,1000.0


In [1643]:
# Imprimimos la forma final del DataFrame limpio, los valores nulos restantes y los tipos de dato finales.
print('Forma final del dataset limpio:', df.shape)
print(' ')
print('Valores nulos restantes:')
print(df.isnull().sum())
print(' ')
print('Tipos de dato finales:')
print(df.dtypes)

Forma final del dataset limpio: (500, 11)
 
Valores nulos restantes:
id_paciente             0
fecha_prescripcion      0
medicamento             0
dosis_mg                0
via_administracion      0
diagnostico_cie10     111
peso_paciente_kg        0
edad                    0
eps                     0
prescriptor_id         42
rango_etario            0
dtype: int64
 
Tipos de dato finales:
id_paciente                    int64
fecha_prescripcion    datetime64[ns]
medicamento                   object
dosis_mg                     float64
via_administracion            object
diagnostico_cie10             object
peso_paciente_kg             float64
edad                           int64
eps                           object
prescriptor_id                object
rango_etario                  object
dtype: object


In [1644]:
# Exportamos el DataFrame limpio a un nuevo archivo CSV.
df.to_csv('prescripciones_limpias_suarez_diaz.csv')
print('Archivo exportado exitosamente: prescripciones_limpias_suarez_diaz.csv')

Archivo exportado exitosamente: prescripciones_limpias_suarez_diaz.csv
